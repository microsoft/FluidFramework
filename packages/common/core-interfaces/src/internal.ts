/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

// eslint-disable-next-line no-restricted-syntax
export * from "./index.js";

// Important: all other exports must be type only exports. In package.json exports,
//  index.js is listed as the runtime file. This is done so that all imports are
//  using the same outer runtime file. (Could be changed if needed.)

// Export set of utility types re-tagged as internal for FF client convenience.
// These types are not intended for direct use by customers and api-extractor will
// flag misuse. If an externally visible version of these types is needed, import
// from via /internal/exposedUtilityTypes rather than /internal.
import type { DeepReadonly as ExposedDeepReadonly } from "./deepReadonly.js";
import type { InternalUtilityTypes as ExposedInternalUtilityTypes } from "./exposedInternalUtilityTypes.js";
import type {
	JsonDeserialized as ExposedJsonDeserialized,
	JsonDeserializedOptions,
} from "./jsonDeserialized.js";
import type {
	JsonSerializable as ExposedJsonSerializable,
	JsonSerializableOptions,
} from "./jsonSerializable.js";
import type {
	JsonTypeWith as ExposedJsonTypeWith,
	ReadonlyNonNullJsonObjectWith as ExposedReadonlyNonNullJsonObjectWith,
} from "./jsonType.js";

// Note: There are no docs for these re-exports. `@inheritdoc` cannot be used as:
//   1. api-extractor does not support renames.
//   2. api-extractor does not support package paths. ("Import paths are not supported")
// Also not useful, at least in VS Code, as substitution is not made in place.

/**
 * @internal
 */
export type DeepReadonly<T> = ExposedDeepReadonly<T>;

/**
 * @internal
 */
export type JsonDeserialized<
	T,
	Options extends JsonDeserializedOptions = {
		AllowExactly: [];
		AllowExtensionOf: never;
	},
> = ExposedJsonDeserialized<T, Options>;

/**
 * @internal
 */
export type JsonSerializable<
	T,
	Options extends JsonSerializableOptions = {
		AllowExactly: [];
		AllowExtensionOf: never;
	},
> = ExposedJsonSerializable<T, Options>;

/**
 * @internal
 */
export type JsonTypeWith<T> = ExposedJsonTypeWith<T>;

/**
 * @internal
 */
export type ReadonlyNonNullJsonObjectWith<T> = ExposedReadonlyNonNullJsonObjectWith<T>;

/**
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace InternalUtilityTypes {
	/* eslint-disable jsdoc/require-jsdoc */
	export type FlattenIntersection<T extends ExposedInternalUtilityTypes.AnyRecord> =
		ExposedInternalUtilityTypes.FlattenIntersection<T>;
	export type IfSameType<
		X,
		Y,
		IfSame = unknown,
		IfDifferent = never,
	> = ExposedInternalUtilityTypes.IfSameType<X, Y, IfSame, IfDifferent>;
	/* eslint-enable jsdoc/require-jsdoc */
}
