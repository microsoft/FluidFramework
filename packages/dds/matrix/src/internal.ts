/*!
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

// Privately exposed to MatrixUndoProvider
export const undoRemoveRows = Symbol("SharedMatrix.undoRemoveRows");
export const undoRemoveCols = Symbol("SharedMatrix.undoRemoveCols");
