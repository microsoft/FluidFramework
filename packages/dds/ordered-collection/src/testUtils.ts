/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

import { ConsensusResult, type IConsensusOrderedCollection } from "./interfaces.js";

/**
 * Helper method to acquire and complete an item
 * Should be used in test code only
 * @internal
 */
export async function acquireAndComplete<T>(
	collection: IConsensusOrderedCollection<T>,
): Promise<T | undefined> {
	let res: T | undefined;
	await collection.acquire(async (value: T) => {
		res = value;
		return ConsensusResult.Complete;
	});
	return res;
}

/**
 * Helper method to acquire and complete an item
 * Should be used in test code only
 * @internal
 */
export async function waitAcquireAndComplete<T>(
	collection: IConsensusOrderedCollection<T>,
): Promise<T> {
	let res: T | undefined;
	await collection.waitAndAcquire(async (value: T) => {
		res = value;
		return ConsensusResult.Complete;
	});
	return res as T;
}

/**
 * Helper method to acquire and release an item
 * Should be used in test code only
 * @internal
 */
export async function acquireAndRelease<T>(
	collection: IConsensusOrderedCollection<T>,
): Promise<T | undefined> {
	let res: T | undefined;
	await collection.acquire(async (value: T) => {
		res = value;
		return ConsensusResult.Release;
	});
	return res;
}

/**
 * Helper method to acquire and release an item
 * Should be used in test code only
 * @internal
 */
export async function waitAcquireAndRelease<T>(
	collection: IConsensusOrderedCollection<T>,
): Promise<T> {
	let res: T | undefined;
	await collection.waitAndAcquire(async (value: T) => {
		res = value;
		return ConsensusResult.Release;
	});
	return res as T;
}
