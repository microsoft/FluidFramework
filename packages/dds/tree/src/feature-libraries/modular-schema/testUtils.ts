/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

import { assert } from "@fluidframework/common-utils";
import { Delta } from "../../tree";

export interface MockChildChange {
    intentions: number[];
    ref: number;
}

export function mockChildChangeRebaser(
    change: MockChildChange,
    baseChange: MockChildChange,
): MockChildChange {
    assert(change.ref === baseChange.ref, "Invalid input passed to child rebaser");
    return {
        intentions: change.intentions,
        ref: baseChange.intentions[baseChange.intentions.length - 1],
    };
}

export function mockChildChangeInverter(change: MockChildChange): MockChildChange {
    return {
        intentions: change.intentions.map((i) => -i).reverse(),
        ref: change.intentions[change.intentions.length - 1],
    };
}

export function mockChildChangeComposer(changes: MockChildChange[]): MockChildChange {
    if (changes.length <= 1) {
        return changes[0];
    }
    const id: number[] = [];
    changes.forEach((change, i) => {
        const prev = changes[i - 1].intentions;
        assert(i === 0 || change.ref === prev[prev.length - 1], "Invalid input to child composer");
        id.push(...change.intentions);
    });
    return {
        intentions: id,
        ref: changes[0].ref,
    };
}

export function mockChildChangeToDelta(change: MockChildChange): Delta.Modify {
    return { type: Delta.MarkType.Modify, setValue: { value: change.intentions.map(String).join("|") } };
}
