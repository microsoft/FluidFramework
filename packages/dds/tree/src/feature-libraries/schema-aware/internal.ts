/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

// Used by put public types, but not part of the desired API surface
export {
	TypeSetToTypedTrees as TreeTypesToTypedTreeTypes,
	TypedSchemaData,
	TypedTree,
	CollectOptions,
	TypedFields,
	ApplyMultiplicity,
	ValueFieldTreeFromSchema,
	FlexibleObject,
} from "./schemaAware";

export { NamesFromSchema, ValuesOf, TypedValue, PrimitiveValueSchema } from "./schemaAwareUtil";
