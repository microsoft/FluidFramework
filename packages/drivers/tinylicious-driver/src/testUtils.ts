/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

export { createInsecureTinyliciousTestTokenProvider } from "./insecureTinyliciousTokenProvider.js";

export {
	createInsecureTinyliciousTestUrlResolver,
	createTinyliciousTestCreateNewRequest,
} from "./insecureTinyliciousUrlResolver.js";
