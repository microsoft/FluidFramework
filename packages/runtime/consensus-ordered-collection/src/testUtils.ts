/*!
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import {
    ConsensusResult,
    IConsensusOrderedCollection,
} from "./interfaces";

/**
 * Helper method to acquire and complete an item
 * Should be used in test code only
 */
export async function acquireAndComplete<T>(collection: IConsensusOrderedCollection<T>): Promise<T | undefined> {
    let res: T | undefined;
    await collection.acquire(async (value: T) => {
        res = value;
        return ConsensusResult.Complete;
    });
    return res;
}

/**
 * Helper method to acquire and complete an item
 * Should be used in test code only
 */
export async function waitAcquireAndComplete<T>(collection: IConsensusOrderedCollection<T>): Promise<T> {
    let res: T | undefined;
    await collection.waitAndAcquire(async (value: T) => {
        res = value;
        return ConsensusResult.Complete;
    });
    return res as T;
}
