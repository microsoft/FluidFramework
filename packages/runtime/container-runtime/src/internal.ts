/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

/* Exports for internal use only
 * This is only used during transition of legacy+alpha code to internal.
 * Here there are conflicts between index reexports with deprecations and clean internal versions.
 * In cases where * exports conflict with named exports, the named exports take precedence per
 * https://tc39.es/ecma262/multipage/ecmascript-language-scripts-and-modules.html#sec-getexportednames.
 * This does trigger the `import/export` lint warning (which is intentionally disabled here). This
 * approach ensures that a non-deprecated version of ContainerRuntime (provided as named direct
 * export) eclipses the deprecated one from `./index.ts`.
 * Alternatively, another file could be created for external exports adding ContainerRuntime and
 * they'd share common exports files (the rest of current index.ts).
 */
/* eslint-disable import/export */
/* eslint-disable import/no-deprecated */

// eslint-disable-next-line no-restricted-syntax
export * from "./index.js";
import { ContainerRuntime as ContainerRuntimeClass } from "./containerRuntime.js";
export type ContainerRuntime = ContainerRuntimeClass;
export const ContainerRuntime = ContainerRuntimeClass;
