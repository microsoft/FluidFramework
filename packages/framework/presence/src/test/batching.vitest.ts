/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

import { EventAndErrorTrackingLogger } from "@fluidframework/test-utils/internal";
import { useFakeTimers, type SinonFakeTimers } from "sinon";
import { describe, it, afterAll, afterEach, beforeAll, beforeEach, expect } from "vitest";

import { Latest, Notifications, type PresenceNotifications } from "../index.js";
import type { createPresenceManager } from "../presenceManager.js";

import { MockRuntimeSignalSnapshotter } from "./snapshotEphemeralRuntime.js";
import { prepareConnectedPresence } from "./testUtils.js";

describe("Presence", () => {
	describe("batching", () => {
		let runtime: MockRuntimeSignalSnapshotter;
		let logger: EventAndErrorTrackingLogger;
		const initialTime = 1000;
		let clock: SinonFakeTimers;
		let presence: ReturnType<typeof createPresenceManager>;

		beforeAll(async () => {
			clock = useFakeTimers();
		});

		beforeEach(() => {
			logger = new EventAndErrorTrackingLogger();
			runtime = new MockRuntimeSignalSnapshotter(logger);
			runtime.snapshotSignals = true;

			// Note that while the initialTime is set to 1000, the prepareConnectedPresence call advances
			// it to 1010 so all tests start at that time.
			clock.setSystemTime(initialTime);

			// Disable submitting signals with a dummy function. This ensures we don't capture signals from
			// test setup, like the prepareConnectedPresence call.
			const submitSignalOriginal = runtime.submitSignal;
			runtime.submitSignal = () => {};

			// Set up the presence connection
			presence = prepareConnectedPresence(runtime, "sessionId-2", "client2", clock, logger);

			// Restore the submiSignal function
			runtime.submitSignal = submitSignalOriginal;
		});

		afterEach(() => {
			clock.reset();
		});

		afterAll(() => {
			clock.restore();
		});

		describe("LatestValueManager", () => {
			it("sends signal immediately when allowable latency is 0", async () => {
				// Configure a state workspace
				const stateWorkspace = presence.getStates("name:testStateWorkspace", {
					count: Latest({ num: 0 }, { allowableUpdateLatencyMs: 0 }),
				}); // SIGNAL #1 - intial data is sent immediately

				const { count } = stateWorkspace.props;

				clock.tick(10); // Time is now 1020

				// SIGNAL #2
				count.local = { num: 42 };

				// SIGNAL #3
				count.local = { num: 84 };

				expect(runtime.submittedSignals).toHaveLength(3);
			});

			it("sets timer for default allowableUpdateLatency", async () => {
				// Configure a state workspace
				presence.getStates("name:testStateWorkspace", {
					count: Latest({ num: 0 } /* default allowableUpdateLatencyMs = 60 */),
				}); // will be queued; deadline is now 1070

				// SIGNAL #1
				// The deadline timer will fire at time 1070 and send a single
				// signal with the value from the last signal (num=0).

				clock.tick(100); // Time is now 1110

				expect(runtime.submittedSignals).toHaveLength(1);
			});

			it("batches signals sent within default allowableUpdateLatency", async () => {
				// Configure a state workspace
				const stateWorkspace = presence.getStates("name:testStateWorkspace", {
					count: Latest({ num: 0 } /* default allowableUpdateLatencyMs = 60 */),
				}); // will be queued; deadline is now 1070

				const { count } = stateWorkspace.props;

				clock.tick(10); // Time is now 1020
				count.local = { num: 12 }; // will be queued; deadline remains 1070

				clock.tick(10); // Time is now 1030
				count.local = { num: 34 }; // will be queued; deadline remains 1070

				clock.tick(30); // Time is now 1060
				count.local = { num: 22 }; // will be queued; deadline remains 1070

				// SIGNAL #1
				// The deadline timer will fire at time 1070 and send a single
				// signal with the value from the last signal (num=22).

				// It's necessary to tick the timer beyond the deadline so the timer will fire.
				clock.tick(20); // Time is now 1080

				clock.tick(10); // Time is now 1090
				count.local = { num: 56 }; // will be queued; deadline is set to 1150

				clock.tick(40); // Time is now 1130
				count.local = { num: 78 }; // will be queued; deadline remains 1150

				clock.tick(10); // Time is now 1140
				count.local = { num: 90 }; // will be queued; deadline remains 1150

				// SIGNAL #2
				// The deadline timer will fire at time 1150 and send a single
				// signal with the value from the last signal (num=90).

				// It's necessary to tick the timer beyond the deadline so the timer will fire.
				clock.tick(30); // Time is now 1180

				expect(runtime.submittedSignals).toHaveLength(2);
			});

			it("batches signals sent within a specified allowableUpdateLatency", async () => {
				// Configure a state workspace
				const stateWorkspace = presence.getStates("name:testStateWorkspace", {
					count: Latest({ num: 0 }, { allowableUpdateLatencyMs: 100 }),
				});

				const { count } = stateWorkspace.props;

				clock.tick(10); // Time is now 1020
				count.local = { num: 12 }; // will be queued; deadline is set to 1120

				clock.tick(80); // Time is now 1100
				count.local = { num: 34 }; // will be queued; deadline remains 1120

				// SIGNAL #1
				// The deadline timer will fire at time 1120 and send a single
				// signal with the value from the last signal (num=34).

				// It's necessary to tick the timer beyond the deadline so the timer will fire.
				clock.tick(30); // Time is now 1130

				clock.tick(10); // Time is now 1140
				count.local = { num: 56 }; // will be queued; deadline is set to 1240

				clock.tick(40); // Time is now 1180
				count.local = { num: 78 }; // will be queued; deadline remains 1240

				clock.tick(40); // Time is now 1220
				count.local = { num: 90 }; // will be queued; deadline remains 1240

				// SIGNAL #2
				// The deadline timer will fire at time 1240 and send a single
				// signal with the value from the last signal (num=90).

				// It's necessary to tick the timer beyond the deadline so the timer will fire.
				clock.tick(30); // Time is now 1250

				expect(runtime.submittedSignals).toHaveLength(2);
			});

			it("queued signal is sent immediately with immediate update message", async () => {
				// Configure a state workspace
				const stateWorkspace = presence.getStates("name:testStateWorkspace", {
					count: Latest({ num: 0 }, { allowableUpdateLatencyMs: 100 }),
					immediateUpdate: Latest({ num: 0 }, { allowableUpdateLatencyMs: 0 }),
				}); // SIGNAL #1 - not queued because it contains a value manager with a latency of 0,
				// so the initial data will be sent immediately

				const { count, immediateUpdate } = stateWorkspace.props;

				clock.tick(10); // Time is now 1020
				count.local = { num: 12 }; // will be queued; deadline is set to 1120

				clock.tick(80); // Time is now 1100
				count.local = { num: 34 }; // will be queued; deadline remains 1120

				clock.tick(10); // Time is now 1110
				immediateUpdate.local = { num: 56 };
				// SIGNAL #2
				// This should cause the queued signals to be merged with this immediately-sent
				// signal with the value from the last signal (num=34).

				// It's necessary to tick the timer beyond the deadline so the timer will fire.
				clock.tick(10); // Time is now 1250

				expect(runtime.submittedSignals).toHaveLength(2);
			});

			it("batches signals with different allowed latencies", async () => {
				// Configure a state workspace
				const stateWorkspace = presence.getStates("name:testStateWorkspace", {
					count: Latest({ num: 0 }, { allowableUpdateLatencyMs: 100 }),
					note: Latest({ message: "" }, { allowableUpdateLatencyMs: 50 }),
				}); // will be queued, deadline is set to 1060

				const { count, note } = stateWorkspace.props;

				clock.tick(10); // Time is now 1020
				note.local = { message: "will be queued" }; // will be queued, deadline remains 1060
				count.local = { num: 12 }; // will be queued; deadline remains 1060

				clock.tick(30); // Time is now 1050
				count.local = { num: 34 }; // will be queued; deadline remains 1060

				clock.tick(10); // Time is now 1060
				note.local = { message: "final message" }; // will be queued; deadline remains 1060

				// SIGNAL #1
				// At time 1060, the deadline timer will fire and send a single signal with the value
				// from the last signal (num=34, message="final message").

				// It's necessary to tick the timer beyond the deadline so the timer will fire.
				clock.tick(30); // Time is now 1080

				expect(runtime.submittedSignals).toHaveLength(1);
			});

			it("batches signals from multiple workspaces", async () => {
				// Configure two state workspaces
				const stateWorkspace = presence.getStates("name:testStateWorkspace", {
					count: Latest({ num: 0 }, { allowableUpdateLatencyMs: 100 }),
				}); // will be queued, deadline is 1110
				const stateWorkspace2 = presence.getStates("name:testStateWorkspace2", {
					note: Latest({ message: "" }, { allowableUpdateLatencyMs: 60 }),
				}); // will be queued, deadline is 1070

				const { count } = stateWorkspace.props;
				const { note } = stateWorkspace2.props;

				clock.tick(10); // Time is now 1020
				note.local = { message: "will be queued" }; // will be queued, deadline is 1080
				count.local = { num: 12 }; // will be queued; deadline remains 1080

				clock.tick(30); // Time is now 1050
				count.local = { num: 34 }; // will be queued; deadline remains 1080

				clock.tick(10); // Time is now 1060
				note.local = { message: "final message" }; // will be queued; deadline remains 1080

				// SIGNAL #1
				// The deadline timer will fire at time 1080 and send a single
				// signal with the values from the last workspace updates (num=34, message="final message").

				// It's necessary to tick the timer beyond the deadline so the timer will fire.
				clock.tick(30); // Time is now 1090

				expect(runtime.submittedSignals).toHaveLength(1);
			});
		});

		describe("NotificationsManager", () => {
			it("notification signals are sent immediately", async () => {
				runtime.snapshotSignals = false;
				// Configure a notifications workspaces
				// eslint-disable-next-line @typescript-eslint/ban-types
				const notificationsWorkspace: PresenceNotifications<{}> = presence.getNotifications(
					"name:testNotificationWorkspace",
					{},
				);

				notificationsWorkspace.add(
					"testEvents",
					Notifications<
						// Below explicit generic specification should not be required.
						{
							newId: (id: number) => void;
						},
						"testEvents"
					>(
						// A default handler is not required
						{},
					),
				);

				const { testEvents } = notificationsWorkspace.props;

				clock.tick(10); // Time is now 1020

				clock.tick(30); // Time is now 1050
				// SIGNAL #1
				testEvents.emit.broadcast("newId", 77);

				expect(
					// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
					(runtime.submittedSignals.at(-1)?.[1] as any).data[
						"n:name:testNotificationWorkspace"
					].testEvents["sessionId-2"].value.args,
				).toEqual([77]);

				clock.tick(10); // Time is now 1060
				// SIGNAL #2
				testEvents.emit.broadcast("newId", 88);

				expect(runtime.submittedSignals).toHaveLength(2);

				expect(
					// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
					(runtime.submittedSignals.at(-1)?.[1] as any).data[
						"n:name:testNotificationWorkspace"
					].testEvents["sessionId-2"].value.args,
				).toEqual([88]);
			});

			it("notification signals cause queued messages to be sent immediately", async () => {
				// Configure a state workspaces
				const stateWorkspace = presence.getStates("name:testStateWorkspace", {
					count: Latest({ num: 0 }, { allowableUpdateLatencyMs: 100 }),
				}); // will be queued, deadline is 1110

				// eslint-disable-next-line @typescript-eslint/ban-types
				const notificationsWorkspace: PresenceNotifications<{}> = presence.getNotifications(
					"name:testNotificationWorkspace",
					{},
				);

				notificationsWorkspace.add(
					"testEvents",
					Notifications<
						// Below explicit generic specification should not be required.
						{
							newId: (id: number) => void;
						},
						"testEvents"
					>(
						// A default handler is not required
						{},
					),
				);

				const { count } = stateWorkspace.props;
				const { testEvents } = notificationsWorkspace.props;

				testEvents.notifications.on("newId", (client, newId) => {
					// do nothing
				});

				clock.tick(10); // Time is now 1020
				count.local = { num: 12 }; // will be queued, deadline remains 1110

				clock.tick(10); // Time is now 1030
				count.local = { num: 34 }; // will be queued, deadline remains 1110

				clock.tick(10); // Time is now 1040
				count.local = { num: 56 }; // will be queued, deadline remains 1110

				clock.tick(20); // Time is now 1060
				testEvents.emit.broadcast("newId", 99);
				// SIGNAL #1
				// The notification will cause an immediate broadcast of the queued signal
				// along with the notification signal.

				clock.tick(30); // Time is now 1090
				// SIGNAL #2
				testEvents.emit.broadcast("newId", 111);

				expect(runtime.submittedSignals).toHaveLength(2);
			});
		});
	});
});
