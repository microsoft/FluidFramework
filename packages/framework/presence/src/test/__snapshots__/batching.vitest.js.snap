// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Presence > batching > LatestValueManager > batches signals from multiple workspaces > submitted signal 1`] = `
"[
  "Pres:DatastoreUpdate",
  {
    "sendTimestamp": 1070,
    "avgLatency": 10,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      },
      "s:name:testStateWorkspace": {
        "count": {
          "sessionId-2": {
            "rev": 2,
            "timestamp": 1050,
            "value": {
              "num": 34
            }
          }
        }
      },
      "s:name:testStateWorkspace2": {
        "note": {
          "sessionId-2": {
            "rev": 2,
            "timestamp": 1060,
            "value": {
              "message": "final message"
            }
          }
        }
      }
    }
  }
]"
`;

exports[`Presence > batching > LatestValueManager > batches signals sent within a specified allowableUpdateLatency > submitted signal 1`] = `
"[
  "Pres:DatastoreUpdate",
  {
    "sendTimestamp": 1110,
    "avgLatency": 10,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      },
      "s:name:testStateWorkspace": {
        "count": {
          "sessionId-2": {
            "rev": 2,
            "timestamp": 1100,
            "value": {
              "num": 34
            }
          }
        }
      }
    }
  }
]"
`;

exports[`Presence > batching > LatestValueManager > batches signals sent within a specified allowableUpdateLatency > submitted signal 2`] = `
"[
  "Pres:DatastoreUpdate",
  {
    "sendTimestamp": 1240,
    "avgLatency": 10,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      },
      "s:name:testStateWorkspace": {
        "count": {
          "sessionId-2": {
            "rev": 5,
            "timestamp": 1220,
            "value": {
              "num": 90
            }
          }
        }
      }
    }
  }
]"
`;

exports[`Presence > batching > LatestValueManager > batches signals sent within default allowableUpdateLatency > submitted signal 1`] = `
"[
  "Pres:DatastoreUpdate",
  {
    "sendTimestamp": 1070,
    "avgLatency": 10,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      },
      "s:name:testStateWorkspace": {
        "count": {
          "sessionId-2": {
            "rev": 3,
            "timestamp": 1060,
            "value": {
              "num": 22
            }
          }
        }
      }
    }
  }
]"
`;

exports[`Presence > batching > LatestValueManager > batches signals sent within default allowableUpdateLatency > submitted signal 2`] = `
"[
  "Pres:DatastoreUpdate",
  {
    "sendTimestamp": 1150,
    "avgLatency": 10,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      },
      "s:name:testStateWorkspace": {
        "count": {
          "sessionId-2": {
            "rev": 6,
            "timestamp": 1140,
            "value": {
              "num": 90
            }
          }
        }
      }
    }
  }
]"
`;

exports[`Presence > batching > LatestValueManager > batches signals with different allowed latencies > submitted signal 1`] = `
"[
  "Pres:DatastoreUpdate",
  {
    "sendTimestamp": 1060,
    "avgLatency": 10,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      },
      "s:name:testStateWorkspace": {
        "count": {
          "sessionId-2": {
            "rev": 2,
            "timestamp": 1050,
            "value": {
              "num": 34
            }
          }
        },
        "note": {
          "sessionId-2": {
            "rev": 1,
            "timestamp": 1020,
            "value": {
              "message": "will be queued"
            }
          }
        }
      }
    }
  }
]"
`;

exports[`Presence > batching > LatestValueManager > queued signal is sent immediately with immediate update message > submitted signal 1`] = `
"[
  "Pres:DatastoreUpdate",
  {
    "sendTimestamp": 1010,
    "avgLatency": 10,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      },
      "s:name:testStateWorkspace": {
        "count": {
          "sessionId-2": {
            "rev": 0,
            "timestamp": 1010,
            "value": {
              "num": 0
            }
          }
        },
        "immediateUpdate": {
          "sessionId-2": {
            "rev": 0,
            "timestamp": 1010,
            "value": {
              "num": 0
            }
          }
        }
      }
    }
  }
]"
`;

exports[`Presence > batching > LatestValueManager > queued signal is sent immediately with immediate update message > submitted signal 2`] = `
"[
  "Pres:DatastoreUpdate",
  {
    "sendTimestamp": 1110,
    "avgLatency": 10,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      },
      "s:name:testStateWorkspace": {
        "count": {
          "sessionId-2": {
            "rev": 2,
            "timestamp": 1100,
            "value": {
              "num": 34
            }
          }
        },
        "immediateUpdate": {
          "sessionId-2": {
            "rev": 1,
            "timestamp": 1110,
            "value": {
              "num": 56
            }
          }
        }
      }
    }
  }
]"
`;

exports[`Presence > batching > LatestValueManager > sends signal immediately when allowable latency is 0 > submitted signal 1`] = `
"[
  "Pres:DatastoreUpdate",
  {
    "sendTimestamp": 1010,
    "avgLatency": 10,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      },
      "s:name:testStateWorkspace": {
        "count": {
          "sessionId-2": {
            "rev": 0,
            "timestamp": 1010,
            "value": {
              "num": 0
            }
          }
        }
      }
    }
  }
]"
`;

exports[`Presence > batching > LatestValueManager > sends signal immediately when allowable latency is 0 > submitted signal 2`] = `
"[
  "Pres:DatastoreUpdate",
  {
    "sendTimestamp": 1020,
    "avgLatency": 10,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      },
      "s:name:testStateWorkspace": {
        "count": {
          "sessionId-2": {
            "rev": 1,
            "timestamp": 1020,
            "value": {
              "num": 42
            }
          }
        }
      }
    }
  }
]"
`;

exports[`Presence > batching > LatestValueManager > sends signal immediately when allowable latency is 0 > submitted signal 3`] = `
"[
  "Pres:DatastoreUpdate",
  {
    "sendTimestamp": 1020,
    "avgLatency": 10,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      },
      "s:name:testStateWorkspace": {
        "count": {
          "sessionId-2": {
            "rev": 2,
            "timestamp": 1020,
            "value": {
              "num": 84
            }
          }
        }
      }
    }
  }
]"
`;

exports[`Presence > batching > LatestValueManager > sets timer for default allowableUpdateLatency > submitted signal 1`] = `
"[
  "Pres:DatastoreUpdate",
  {
    "sendTimestamp": 1070,
    "avgLatency": 10,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      },
      "s:name:testStateWorkspace": {
        "count": {
          "sessionId-2": {
            "rev": 0,
            "timestamp": 1010,
            "value": {
              "num": 0
            }
          }
        }
      }
    }
  }
]"
`;

exports[`Presence > batching > NotificationsManager > notification signals cause queued messages to be sent immediately > submitted signal 1`] = `
"[
  "Pres:DatastoreUpdate",
  {
    "sendTimestamp": 1060,
    "avgLatency": 10,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      },
      "s:name:testStateWorkspace": {
        "count": {
          "sessionId-2": {
            "rev": 3,
            "timestamp": 1040,
            "value": {
              "num": 56
            }
          }
        }
      },
      "n:name:testNotificationWorkspace": {
        "testEvents": {
          "sessionId-2": {
            "rev": 0,
            "timestamp": 1060,
            "value": {
              "name": "newId",
              "args": [
                99
              ]
            },
            "ignoreUnmonitored": true
          }
        }
      }
    }
  }
]"
`;

exports[`Presence > batching > NotificationsManager > notification signals cause queued messages to be sent immediately > submitted signal 2`] = `
"[
  "Pres:DatastoreUpdate",
  {
    "sendTimestamp": 1090,
    "avgLatency": 10,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      },
      "n:name:testNotificationWorkspace": {
        "testEvents": {
          "sessionId-2": {
            "rev": 0,
            "timestamp": 1090,
            "value": {
              "name": "newId",
              "args": [
                111
              ]
            },
            "ignoreUnmonitored": true
          }
        }
      }
    }
  }
]"
`;
